test/src/e2e.dynval/dv.n-1.test.ts

```diff
@@ -0,0 +1,184 @@
+// This file is part of CodeChain.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+import * as chai from "chai";
+import { expect } from "chai";
+import * as chaiAsPromised from "chai-as-promised";
+import { SDK } from "codechain-sdk";
+import * as stake from "codechain-stakeholder-sdk";
+import "mocha";
+
+import { validators as originalValidators } from "../../tendermint.dynval/constants";
+import { faucetAddress, faucetSecret } from "../helper/constants";
+import { PromiseExpect } from "../helper/promise";
+import { withNodes } from "./setup";
+import CodeChain from "../helper/spawn";
+
+async function getPossibleAuthors(
+    sdk: SDK,
+    blockNumber: number
+): Promise<null | string[]> {
+    return new Promise((resolve, reject) => {
+        sdk.rpc
+            .sendRpcRequest("chain_getPossibleAuthors", [blockNumber])
+            .then(result => {
+                if (
+                    result === null ||
+                    result.every((str: any) => typeof str === "string")
+                ) {
+                    resolve(result);
+                }
+                return reject(
+                    Error(
+                        `Expected chain_getPossibleAuthors to return an array of string or null but it returned ${result}`
+                    )
+                );
+            })
+            .catch(reject);
+    });
+}
+
+chai.use(chaiAsPromised);
+
+const [alice, , , ...otherDynValidators] = originalValidators;
+const allDynValidators = [alice, ...otherDynValidators];
+
+describe("Dynamic Validator N -> N-1", function() {
+    this.timeout(0);
+
+    const promiseExpect = new PromiseExpect();
+    const termSeconds = 20;
+
+    const nodes = withNodes(this, {
+        promiseExpect,
+        validators: allDynValidators.map((signer, index) => ({
+            signer,
+            delegation: 5_000,
+            deposit: 10_000_000 - index // tie-breaker
+        })),
+        overrideParams: {
+            termSeconds,
+            maxNumOfValidators: 8
+        }
+    });
+
+    describe("A node is imprisoned to jail", async function() {
+        it("alice should be dropped out from validator list", async function() {
+            const margin = 1.2;
+            const aliceNode = nodes[0];
+            const checkingNode = nodes[1];
+
+            await aliceNode.clean();
```

Comment:
Alice is cleaned after the first term is started. It might lead to a flaky test. (What if Alice is the first validator and, there's a race between `clean()` and the term's first block? There are only 5 seconds to do `clean()`). This might be why you've left a comment `// not to make timing issue, wait until the term 3.`, and this is why we have `onBeforeEnable` in `withNode`'s parameter. 

Can you move `const nodes = withNodes(...)` to this inner `describe` from outer `describe`, and add `clean()` to `onBeforeEnable` for this case? 
It'll remove unnecessary waiting for term 3. 
You can share parameters like this to avoid duplications: https://github.com/CodeChain-io/codechain/pull/1690/files#diff-d5a0bb0ab12e8c7ae2dfbe269d53dceeR44. 
Duplicating it might also acceptable because it is not quite long.

Reply:
I Introduced `onBeforeEnable`, thanks.